### ISSUE TITLE
Are Aliases really necessary for GraphQL?

### ISSUE THREAD

**Issue by Pinski** [Yelp/yelp-fusion#231]
Hello,

So personally, I don't see the use case for the aliases in the GraphQL if I request multiple businesses.  There might be other use cases out there that are not for me and I get that.  But currently, the aliases just get in my way.  

When I use the search endpoint, the framework that I use (C#) can easily cast that JSON result to a list of models.  However, with the extra text in the GraphQL request, this is more difficult.  To use the GraphQL I currently need to make a request, get the JSON string back, basically do string manipulations to strip out all of the aliases, this makes it look like an array of businesses just like the search endpoint, and then I can cast that to a list of models.  The only other way I can think of is using dynamic objects or models that contain hardcoded versions of the alias names as object names and I don't want to do that either.

I guess I was expecting to be able to send an array of ids with the same fragment and get the same type of results I would get back from search.

**Request Body**
```
{
    business(id: ["fiero-caffe-san-mateo-2", "ninis-coffee-shop-san-mateo"]) { 
		id
		name
		rating
		price
    }
}
```
or
```
{
    business(id: "fiero-caffe-san-mateo-2") {
        ...basicBizInfo
    },
    business(id: "ninis-coffee-shop-san-mateo") {
        ...basicBizInfo
    }
}

fragment basicBizInfo on Business {
	id
	name
	rating
	price
}
```
**Response Body**
```
{
  "businesses": [
    {
      "id": "fiero-caffe-san-mateo-2",
      "name": "Fiero Caffe",
      "rating": 4.5,
      "price": "$$"
    },
    {
      "id": "ninis-coffee-shop-san-mateo",
      "name": "Nini's Coffee Shop",
      "rating": 4.5,
      "price": "$$"
    }
  ]
}
```

This seems a lot more straight forward to me.  Do you think the aliases could become optional in the future?  Or maybe the alias could become part of the business data model?
```
{
      "alias": "b1",
      "id": "ninis-coffee-shop-san-mateo",
      "name": "Nini's Coffee Shop",
      "rating": 4.5,
      "price": "$$"
 }
```

What do you think?  Do my suggestions go against what you are trying to achieve?  Am I missing something?

The GraphQL endpoint is definitely something I want to utilize as I have lists of businesses that I want to retrieve in as few calls as possible.  I'm just curious about your thoughts on the matter going forward as I try to decide how much engineering effort to put forth in making clean GraphQL implementations on my side.  Definitely not expecting timelines or promises.  Just curious what your thoughts are.

Thanks.

**Comment by mittonk** [comment]
@tomelm : Everything I know I about this I learned from http://graphql.org/learn/queries/#aliases , got any more targeted suggestions?

**Comment by tomelm** [comment]
@Pinski aliases are the GraphQL way of doing bulk requests. An array parameter would work but it would be harder for us to guarantee the order the order of results due to how dataloading works or you'd have to go through the response and sort the values yourself.

As far as bulk requests without aliases (your second example): that won't work in GraphQL since each query field maps to an object response.

I might not fully understand how C# works but why do you need to do string manipulation to parse out the alias?

If you had a query that was
```
{
  b1: business(...) {...}
  b2: business(...) {...}
}
```
couldn't you index into that key in the response? `data['b1']` and `data['b2']`?

**Comment by Pinski** [comment]
Hi @mittonk and @tomelm,

Thanks for taking the time to respond to me.  This is my introduction to GraphQL and I find it interesting.  I wouldn't presume to tell you how to change your API, just curious and giving you my experience with it.

Personally, I have no need to have the responses sorted in any type of order.  The most ordering I think I could ever possibly want is all the businesses to be in alphabetic order.  Again, I realize there might be other use cases besides mine.

I might be able to index into the response easier if it was an array.  However, the response that I get right now is a JSON nested object, not a JSON array (unless I'm missing something).  This is the response I'm receiving:
```
{
  "data": {
    "b1": {
      "id": "yakitori-kokko-san-mateo",
      "photos": ["https: //s3-media3.fl.yelpcdn.com/bphoto/QGAPGKvIxbELnad67m0J4Q/o.jpg"],
      "name": "YakitoriKokko",
      "url": "https: //www.yelp.com/biz/yakitori-kokko-san-mateo?adjust_creative=O_9X4ahUd0ypRrt7Bx8Z5A&utm_campaign=yelp_api_v3&utm_medium=api_v3_graphql&utm_source=O_9X4ahUd0ypRrt7Bx8Z5A",
      "rating": 4.0,
      "review_count": 739,
      "price": "$$",
      "categories": [{
        "title": "Japanese",
        "alias": "japanese"
      }],
      "location": {
        "address1": "5092ndAve",
        "address2": "",
        "address3": "",
        "city": "SanMateo",
        "state": "CA",
        "zip_code": "94401"
      },
      "display_phone": "(650)401-7008",
      "coordinates": {
        "latitude": 37.568120075232,
        "longitude": -122.321434020996
      },
      "hours": [{
        "is_open_now": false
      }]
    },
    "b2": {
      "id": "zalla-kabab-house-danville",
      "photos": ["https: //s3-media4.fl.yelpcdn.com/bphoto/9c6lOnHBi1mJaXrCvETVYQ/o.jpg"],
      "name": "ZallaKababHouse",
      "url": "https: //www.yelp.com/biz/zalla-kabab-house-danville?adjust_creative=O_9X4ahUd0ypRrt7Bx8Z5A&utm_campaign=yelp_api_v3&utm_medium=api_v3_graphql&utm_source=O_9X4ahUd0ypRrt7Bx8Z5A",
      "rating": 4.5,
      "review_count": 97,
      "price": "$$",
      "categories": [{
        "title": "Mediterranean",
        "alias": "mediterranean"
      },
      {
        "title": "Afghan",
        "alias": "afghani"
      }],
      "location": {
        "address1": "202SycamoreValleyRdW",
        "address2": null,
        "address3": "",
        "city": "Danville",
        "state": "CA",
        "zip_code": "94526"
      },
      "display_phone": "(925)208-1403",
      "coordinates": {
        "latitude": 37.81221,
        "longitude": -121.99733
      },
      "hours": [{
        "is_open_now": false
      }]
    }
  }
}
```

C# can only convert this into a concrete object.  I've tried many ways of converting into some dynamic type and C# and Json.NET just return nulls.  Unless I'm missing something in the C# language, the only way to convert this would be to have an actual class like the following:
```
public class Data
{
	public BusinessResponse B1 { get; set; }
	public BusinessResponse B2 { get; set; }
	...
}
```

You can see how this would be problematic, I'd have to hardcode and define all the expected aliases as part of the concrete class.  Then, after casting the JSON to the class, I'd have to programmatically go through the class and check for nulls, use the values that aren't nulls to build an Array / List, etc.

So what I'm currently doing is string manipulation to turn this into an array of BusinessResponses (my concrete class for a Yelp Business response):
```
[{
  "id": "yakitori-kokko-san-mateo",
  "photos": ["https://s3-media3.fl.yelpcdn.com/bphoto/QGAPGKvIxbELnad67m0J4Q/o.jpg"],
  "name": "Yakitori Kokko",
  "url": "https://www.yelp.com/biz/yakitori-kokko-san-mateo?adjust_creative=O_9X4ahUd0ypRrt7Bx8Z5A&utm_campaign=yelp_api_v3&utm_medium=api_v3_graphql&utm_source=O_9X4ahUd0ypRrt7Bx8Z5A",
  "rating": 4.0,
  "review_count": 739,
  "price": "$$",
  "categories": [{
    "title": "Japanese",
    "alias": "japanese"
  }],
  "location": {
    "address1": "509 2nd Ave",
    "address2": "",
    "address3": "",
    "city": "San Mateo",
    "state": "CA",
    "zip_code": "94401"
  },
  "display_phone": "(650) 401-7008",
  "coordinates": {
    "latitude": 37.568120075232,
    "longitude": -122.321434020996
  },
  "hours": [{
    "is_open_now": false
  }]
},
{
  "id": "zalla-kabab-house-danville",
  "photos": ["https://s3-media4.fl.yelpcdn.com/bphoto/9c6lOnHBi1mJaXrCvETVYQ/o.jpg"],
  "name": "Zalla Kabab House",
  "url": "https://www.yelp.com/biz/zalla-kabab-house-danville?adjust_creative=O_9X4ahUd0ypRrt7Bx8Z5A&utm_campaign=yelp_api_v3&utm_medium=api_v3_graphql&utm_source=O_9X4ahUd0ypRrt7Bx8Z5A",
  "rating": 4.5,
  "review_count": 97,
  "price": "$$",
  "categories": [{
    "title": "Mediterranean",
    "alias": "mediterranean"
  },
  {
    "title": "Afghan",
    "alias": "afghani"
  }],
  "location": {
    "address1": "202 Sycamore Valley Rd W",
    "address2": null,
    "address3": "",
    "city": "Danville",
    "state": "CA",
    "zip_code": "94526"
  },
  "display_phone": "(925) 208-1403",
  "coordinates": {
    "latitude": 37.81221,
    "longitude": -121.99733
  },
  "hours": [{
    "is_open_now": false
  }]
}]
```

This is easily translated by C# into an Array / List / IEnumerable of BusinessResponses.

As far as I can tell, there's no way to make an Associative Array in JSON.  So that's why I was suggesting the alias becomes part of the object.  That way you can have a JSON array of objects without the alias being in the way.

```
[{
	"alias": "b1",
	"id": "ninis-coffee-shop-san-mateo",
	"name": "Nini's Coffee Shop",
	"rating": 4.5,
	"price": "$$"
},
{
	"alias": "b2",
	"id": "zalla-kabab-house-danville",
	"name": "Zalla Kabab House",
	"rating": 4.5,
	"price": "$$",
}]
```

If this breaks what you are trying to accomplish or messes up other use cases that I'm not seeing, that's fine.  I have my string manipulation solution that works perfectly for now.  It's just less than ideal in my opinion.

Thanks again.

**Comment by tomelm** [comment]
Ahh, I remember this problem a bit from when I used to do Java. Unfortunately I don't know enough C# to help out here but generally some json libraries do have ways to handle polymorphic endpoints with changing keys.

Returning a list here as you proposed would require some weird changes to how we construct the responses and wouldn't be following the spec which I'm worried will cause problems with some of the existing graphql client libraries. Sorry about the difficulty @Pinski!

**Comment by Pinski** [comment]
Hey, no worries.  I appreciate you getting back to me @tomelm.  I'm happy to know what you are thinking so I can figure out how much or little I want to change what I'm working on.  Didn't want to spend a sizeable amount of time fixing it and find out a patch came out the next day.

What I'm doing now works perfectly well and seems to be pretty quick; even if it's a little messy doing string replacements.  I might investigate a way to get Json.NET to handle it in the future if it becomes a problem.

Thanks again.

**Comment by watterso** [comment]
@Pinski I think I might have solution :)

I modified the code from [this bit of the JSON.Net docs](http://www.newtonsoft.com/json/help/html/SerializingCollections.htm#DeserializingDictionaries):
```c#
string json; //Your JSON response
public class BusinessResponse {} //I assume this is your class?
Dictionary<string, BusinessResponse> responseDictionary = JsonConvert.DeserializeObject<Dictionary<string, BusinessResponse>>(json);
// Do things with responseDictionary
// Convert it to list of BusinessResponse's maybe?
```

I didn't run this code and my C# is also rusty, but hopefully its a helpful starting point :)

**Comment by Pinski** [comment]
@watterso for some reason I didn't even consider mapping that to a dictionary.  I think I was focused too much on Arrays and Lists.  Unfortunately, it didn't convert directly to a dictionary either.  

But then I realized since the GraphQL response is a standard, there must be a .NET implementation.  That lead me to a few projects and examples.  I learned more about the Newtonsoft.Json.Linq.JObject type, which I never heard of before.  Between that type and your suggestion about dictionaries, I finally got it to work.  
Here it is in case anyone else needs it in the future.

```
var jObject = JObject.Parse(jsonResponse);
var businessResponseDictionary = jObject["data"].ToObject<Dictionary<string, BusinessResponse>>();
var businessResponseList = businessResponseDictionary.Select(keyValuePair => keyValuePair.Value).ToList();
```

Thanks for all the help and suggestions everyone.
